### PR TITLE
Time service task finishing

### PR DESCRIPTION
**Pull request by wojtas626**

### CODE REVIEW

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 1]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
   }
 
   public void start(Task task) {
+    if (isValidNotStartedTask(task)) {
+      throw new IllegalArgumentException();
+    }
     task.setStart(LocalDateTime.now(clock));
   }
+
+  private boolean isValidNotStartedTask(Task task) {
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: final Task

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 2]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
...
+    return task.getStart() != null
+        || task.getFinish() != null;
+  }
+
+  public void finish(Task task) {
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: final Task

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 3]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
...
+    LocalDateTime finishTime = LocalDateTime.now(clock);
+    if (isValidStartedTask(task, finishTime)) {
+      throw new IllegalArgumentException();
+    }
+    task.setFinish(finishTime);
+  }
+
+  private boolean isValidStartedTask(Task task, LocalDateTime finishTime) {
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: final Task, final LocalDateTime

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 4]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
...
   public void start(Task task) {
+    if (isValidNotStartedTask(task)) {
+      throw new IllegalArgumentException();
+    }
     task.setStart(LocalDateTime.now(clock));
   }
+
+  private boolean isValidNotStartedTask(Task task) {
+    return task.getStart() != null
+        || task.getFinish() != null;
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: rather and(&&) condition

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 5]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
...
+  }
+
+  public void finish(Task task) {
+    LocalDateTime finishTime = LocalDateTime.now(clock);
+    if (isValidStartedTask(task, finishTime)) {
+      throw new IllegalArgumentException();
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: Why you throw the exception when the task is valid.

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 6]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
...
+    }
+    task.setFinish(finishTime);
+  }
+
+  private boolean isValidStartedTask(Task task, LocalDateTime finishTime) {
+    return task.getStart() == null
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: why task.getStart() == null?

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java (hunk 4 above)
Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: ???

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java (hunk 1 above)
Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: too verbose

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java (hunk 3 above)
Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: too verbose

Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: (effectively final)

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java (hunk 5 above)
Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: method should be renamed only

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java (hunk 6 above)
Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: method should be renamed only

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java (hunk 2 above)
Comment (wojtas626) [reply src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: too verbose (effectively final)

File: src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java [hunk 7]
```diff
@@ -13,6 +13,28 @@ public TimeService(Clock clock) {
   }
 
   public void start(Task task) {
+    if (isInvalidValidNotStartedTask(task)) {
+      throw new IllegalArgumentException();
+    }
     task.setStart(LocalDateTime.now(clock));
   }
+
+  private boolean isInvalidValidNotStartedTask(Task task) {
```
Comment (matkluska) [inline src/main/java/pl/edu/agh/kis/timetracker/service/TimeService.java]: C`mon

**Pull request merged**